File: vps-renew-notification-go/internal/db/vps.go
package db

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/axumrs/vps-renew-notification-go/internal/filter"
	"github.com/axumrs/vps-renew-notification-go/internal/model"
	"github.com/axumrs/vps-renew-notification-go/internal/pkg"
	"github.com/jmoiron/sqlx"
)

// CreateVps 创建VPS
func CreateVps(ctx context.Context, q sqlx.ExecerContext, m *model.VPS) (string, error) {
	id := pkg.NewID()
	if _, err := q.ExecContext(ctx, "INSERT INTO vpss (id, provider_id, name, expire, dateline) VALUES ($1, $2, $3, $4, $5)", id, m.ProviderID, m.Name, time.Time(m.Expire).In(time.Local), m.Dateline); err != nil {
		return "", err
	}
	return id, nil
}

// UpdateVps 更新Vps
func UpdateVps(ctx context.Context, q sqlx.ExecerContext, m *model.VPS) (int64, error) {
	r, err := q.ExecContext(ctx, "UPDATE vpss SET  provider_id=$1, name=$2, expire=$3 WHERE id=$4", m.ProviderID, m.Name, time.Time(m.Expire).In(time.Local), m.ID)
	if err != nil {
		return 0, err
	}
	return r.RowsAffected()
}

// DeleteVps 删除VPS
func DeleteVps(ctx context.Context, q sqlx.ExecerContext, id string) (int64, error) {
	return Delete(ctx, q, "vpss", id)
}

// FindVps 查找VPS
func FindVps(ctx context.Context, q sqlx.QueryerContext, id string) (*model.VPS, error) {
	var m model.VPS
	if err := sqlx.GetContext(ctx, q, &m, "SELECT id, provider_id, name, expire, dateline FROM vpss WHERE id=$1", id); err != nil {
		return nil, err
	}
	return &m, nil
}

// ListVps VPS列表
func ListVps(ctx context.Context, q sqlx.QueryerContext, f *filter.VpsListFilter) ([]*model.VPSWithProvider, error) {
	ls := []*model.VPSWithProvider{}

	params := []any{}
	sb := strings.Builder{}
	sb.WriteString("SELECT id, provider_id, name, expire, dateline,provider_name FROM v_vps_proiders WHERE 1=1")

	if !pkg.IsEmptyStr(f.Name) {
		sb.WriteString(fmt.Sprintf(" AND name ILIKE $%d", len(params)+1))
		params = append(params, fmt.Sprintf("%%%s%%", f.Name))
	}
	if !pkg.IsEmptyStr(f.ProviderID) {
		sb.WriteString(fmt.Sprintf(" AND provider_id = $%d", len(params)+1))
		params = append(params, f.ProviderID)
	}
	if !pkg.IsEmptyStr(f.Sort) {
		sb.WriteString(" ORDER BY ")
		sb.WriteString(f.Sort)
	} else {
		sb.WriteString(" ORDER BY id DESC")
	}

	if err := sqlx.SelectContext(ctx, q, &ls, sb.String(), params...); err != nil {
		return nil, err
	}

	return ls, nil
}

File: vps-renew-notification/src/db/vps.rs
use sqlx::{query, query_as, PgExecutor, QueryBuilder, Result};

use crate::{filter, id, model};

pub async fn create<'a>(c: impl PgExecutor<'a>, m: &'a model::VPS) -> Result<String> {
    let id = id::new();

    query("INSERT INTO vpss (id, provider_id, name, expire, dateline) VALUES ($1, $2, $3, $4, $5)")
        .bind(&id)
        .bind(&m.provider_id)
        .bind(&m.name)
        .bind(&m.expire)
        .bind(&m.dateline)
        .execute(c)
        .await?;

    Ok(id)
}

pub async fn update<'a>(c: impl PgExecutor<'a>, m: &'a model::VPS) -> Result<u64> {
    let r = query("UPDATE vpss SET  provider_id=$1, name=$2, expire=$3 WHERE id=$4")
        .bind(&m.provider_id)
        .bind(&m.name)
        .bind(&m.expire)
        .bind(&m.id)
        .execute(c)
        .await?;
    Ok(r.rows_affected())
}

pub async fn delete<'a>(c: impl PgExecutor<'a>, id: &'a str) -> Result<u64> {
    super::delete(c, "vpss", id).await
}

pub async fn find<'a>(c: impl PgExecutor<'a>, id: &'a str) -> Result<Option<model::VPS>> {
    let m = query_as("SELECT id, provider_id, name, expire, dateline FROM vpss WHERE id=$1")
        .bind(id)
        .fetch_optional(c)
        .await?;
    Ok(m)
}

pub async fn list<'a>(
    c: impl PgExecutor<'a>,
    f: &'a filter::VpsListFilter,
) -> Result<Vec<model::VPSWithProvider>> {
    let mut q =
        QueryBuilder::new("SELECT id, provider_id, name, expire, dateline,provider_name,renew_days,notify_days FROM v_vps_proiders WHERE 1=1");
    if let Some(name) = &f.name {
        q.push(" AND name ILIKE ").push_bind(format!("%{name}%"));
    }

    if let Some(provider_id) = &f.provider_id {
        q.push(" AND provider_id=").push_bind(provider_id);
    }

    if let Some(sort) = &f.sort {
        q.push(format!(" ORDER BY {sort}"));
    } else {
        q.push(" ORDER BY id DESC");
    }

    let ls = q.build_query_as().fetch_all(c).await?;

    Ok(ls)
}

pub async fn batch_find<'a>(
    c: impl PgExecutor<'a>,
    ids: &[&str],
) -> Result<Vec<model::VPSWithProvider>> {
    let mut q =
        QueryBuilder::new("SELECT id, provider_id, name, expire, dateline,provider_name,renew_days,notify_days FROM v_vps_proiders WHERE id IN");
    q.push_tuples(ids, |mut b, id| {
        b.push_bind(id);
    });
    q.build_query_as().fetch_all(c).await
}
